### DISCUSSION TITLE
GPIO Interrupt?

### DISCUSSION THREAD

**Discussion by lazyoldjack** [toitlang/toit#760] [Q&A]
Is it planned to handle GPIO interrupts, or have I missed something when looking for such a feature?

**Comment by kasperl** [comment]
@lazyoldjack You can use [gpio.Pin.wait_for](https://libs.toit.io/gpio/class-Pin#wait_for(1%2C0%2C0%2C)) which uses interrupts underneath.

**Reply by lazyoldjack** [reply]
I had tried this but couldn't get over about 30Hz; probably my implementation. The pulse counter (below) might well be the answer. Thanks!

**Comment by floitsch** [comment]
Often the ESP32 peripherals can do tasks that you would otherwise do in the interrupts.
Can you give me an example of where you would want to use the GPIO interrupt?

**Reply by lazyoldjack** [reply]
I wanted to measure the frequency of a square wave on a GPIO pin, up to about 2Khz, and I want to measure it about 10 times a second, to catch variance.

**Reply by floitsch** [reply]
I committed the pulse counter library today: https://github.com/toitlang/toit/blob/master/lib/pulse_counter.toit

Maybe that would work for your use case?

**Reply by lazyoldjack** [reply]
Wow, I'd forgotten that the ESP32 had such a thing. I guess I should re-read the specs from time to time!
Yes, this looks promising. 'Scuse my ignorance, but will this appear as a package, or in an updated version of Toit firmware? I did look for a package but couldn't find it...

**Reply by floitsch** [reply]
It's already available as a built-in library on master.
As soon as we release a new version it will also be available in the prebuilt releases and in Jaguar.
